### ISSUE TITLE
Rewrite resources to cdnjs

### ISSUE THREAD

**Issue by apeschar** [kiboit/phast#57]
Many resources (scripts, stylesheets and fonts) on most sites are libraries that are also available on public CDNs. [cdnjs](https://cdnjs.com/) has the highest coverage and supports subresource integrity (SRI).

I am not aware of a public list of all the files and their hashes that are hosted on cdnjs. They provide hashes for scripts and stylesheets, but not images and fonts.

Using a clone of their repo, we should be able to generate that list ourselves. Then we can expose it using a DNS server or HTTPS API. Phast can check the hashes of local resources against that server and rewrite their URLs, if the resource is found on cdnjs.

(I'm currently cloning the cdnjs repo to see if I can set up a DNS resolver that maps hashes to URLs.)

**Comment by apeschar** [comment]
In order to achieve this, I need to:

* Set up PowerDNS with a PostgreSQL back-end
* Write a script to create records based on the cdnjs repo contents
* Set up a cronjob for that script

Then, we can `dig TXT abcdef.cdnjs.phastcdn.net` and get a URL if the file is on cdnjs.

**Comment by MartinKolarik** [comment]
You might want to checkout [jsDelivr CDN](https://www.jsdelivr.com/). It hosts all projects from npm and GitHub and also offers an [API](https://github.com/jsdelivr/data.jsdelivr.com) (which provides SRI hashes for all files, among other things).

**Comment by apeschar** [comment]
Thanks for the tip.

The challenge is to map file hashes to file URLs on the CDN.

cdnjs seemed easier because they have a big repo for which we could quickly write [a script to index all the files by hash](https://github.com/kiboit/cdnjs-lookup/blob/master/cdnjs2pdns.py).

jsDelivr has more stuff, but we'd have to crawl the API to build a database of all the hashes and their URLs. That's slightly more work. We might do that in a later iteration.

~~~
% dig +short txt 3cvcj3ggz3fruycrlpajh4oj3i4kaojgclm2xcxa67zw43xod6wq.cdnjs.phastcdn.net
"https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.js"
~~~

**Comment by MartinKolarik** [comment]
Indeed, I missed the part about mapping files by their hash. I think I could help you with that though - I'm from jsDelivr and I find this idea is really interesting.

We should be able to extend our API to allow lookups by file hash, so something like `GET /v1/find-by-hash/sha256` would give you a CDN url if it exists. It would work for any package, which was previously accessed from our CDN at least once, and would be updated automatically as new packages are released.

I think this would be much easier for you to work with, as you wouldn't need to maintain a copy of the repo, crawl it, and maintain a DB. What do you think?
